src/Microsoft.Diagnostics.DebugServices.Implementation/ServiceManager.cs

```diff
@@ -232,30 +311,79 @@ public void LoadExtension(string extensionPath)
         }
 
         /// <summary>
-        /// Register the exported services in the assembly and notify the assembly has loaded.
+        /// Load the extension using an assembly load context. This needs to be in
+        /// a separate method so ExtensionLoadContext class doesn't get referenced
+        /// when running on desktop Framework.
         /// </summary>
-        /// <param name="assembly">extension assembly</param>
-        public void RegisterAssembly(Assembly assembly)
+        /// <param name="extensionPath">extension assembly path</param>
+        /// <returns>assembly</returns>
+        private Assembly UseAssemblyLoadContext(string extensionPath)
         {
-            if (_finalized)
+            ExtensionLoadContext extension = new(extensionPath);
+            Assembly assembly = extension.LoadFromAssemblyPath(extensionPath);
+            if (assembly is not null)
             {
-                throw new InvalidOperationException();
+                // This list is just to keep the load context alive
+                _extensions.Add(extension);
             }
+            return assembly;
+        }
 
-            try
+        private sealed class ExtensionLoadContext : AssemblyLoadContext
+        {
+            private static readonly HashSet<string> s_defaultAssemblies = new() {
+                "Microsoft.Diagnostics.DebugServices",
+                "Microsoft.Diagnostics.DebugServices.Implementation",
+                "Microsoft.Diagnostics.ExtensionCommands",
+                "Microsoft.Diagnostics.NETCore.Client",
+                "Microsoft.Diagnostics.Repl",
+                "Microsoft.Diagnostics.Runtime",
+                "Microsoft.FileFormats",
+                "Microsoft.SymbolStore",
+                "SOS.Extensions",
+                "SOS.Hosting",
+                "SOS.InstallHelper"
+            };
+
+            private Dictionary<string, string> _extensionPaths;
+
+            public string ExtensionPath { get; }
+
+            public ExtensionLoadContext(string extensionPath)
             {
-                RegisterExportedServices(assembly);
-                _notifyExtensionLoad.Fire(assembly);
+                ExtensionPath = extensionPath;
             }
-            catch (Exception ex) when (ex is DiagnosticsException or NotSupportedException or FileNotFoundException)
+
+            protected override Assembly Load(AssemblyName assemblyName)
             {
-                Trace.TraceError(ex.ToString());
+                lock (this)
+                {
+                    if (_extensionPaths == null)
+                    {
+                        string[] extensionFiles = Directory.GetFiles(Path.GetDirectoryName(ExtensionPath), "*.dll");
+                        _extensionPaths = new Dictionary<string, string>();
```

Comment:
I wonder if `AssemblyDependencyResolver` is more appropriate here. Also, should this be done in the constructor/init/factory instead of initializing this every time?

Reply:
If not, is it worth it doing this? Is it better to cache `Path.GetDirectoryName(ExtensionPath)` and then just do 
```
string extDepPath = Path.Combine(folder, assemblyName.Name);
return File.Exists(extDepPath) ?? LoadFromAssemblyPath(extDepPath) : null;
```

after the `s_defaultAssemblies.Contains(assemblyName.Name)` check.

Reply:
I still don't know if we need a dictionary for this rather than building the path and checking once.